s3/src/main/java/ch/cyberduck/core/sts/STSAssumeRoleTokenExpiredResponseInterceptor.java

```diff
@@ -0,0 +1,109 @@
+
+    private static final int MAX_RETRIES = 1;
+
+    private final S3Session session;
+    private final OAuth2RequestInterceptor oauth;
+    private final STSAssumeRoleCredentialsRequestInterceptor sts;
+
+    public STSAssumeRoleTokenExpiredResponseInterceptor(final S3Session session,
+                                                        final OAuth2RequestInterceptor oauth,
+                                                        final STSAssumeRoleCredentialsRequestInterceptor sts,
+                                                        final LoginCallback prompt) {
+        super(session.getHost(), oauth, prompt);
+        this.session = session;
+        this.oauth = oauth;
+        this.sts = sts;
+    }
+
+    @Override
+    public boolean retryRequest(final HttpResponse response, final int executionCount, final HttpContext context) {
+        switch(response.getStatusLine().getStatusCode()) {
+            case HttpStatus.SC_UNAUTHORIZED:
+                if(!super.retryRequest(response, executionCount, context)) {
+                    return false;
+                }
+        }
+        switch(response.getStatusLine().getStatusCode()) {
+            case HttpStatus.SC_BAD_REQUEST:
+                try {
+                    if(null != response.getEntity()) {
+                        EntityUtils.updateEntity(response, new BufferedHttpEntity(response.getEntity()));
+                        final S3ServiceException failure = new S3ServiceException(response.getStatusLine().getReasonPhrase(),
+                                EntityUtils.toString(response.getEntity()));
+                        failure.setResponseCode(response.getStatusLine().getStatusCode());
+                        if(new S3ExceptionMappingService().map(failure) instanceof ExpiredTokenException) {
+                            if(log.isWarnEnabled()) {
+                                log.warn(String.format("Handle failure %s", failure));
+                            }
+                        }
+                        else {
+                            // Ignore other 400 failures
+                            return false;
+                        }
+                    }
+                }
+                catch(IOException e) {
+                    log.warn(String.format("Failure parsing response entity from %s", response));
+                }
+                // Break through
+            case HttpStatus.SC_UNAUTHORIZED:
+                if(executionCount <= MAX_RETRIES) {
+                    try {
+                        log.warn(String.format("Attempt to refresh STS token for failure %s", response));
+                        final STSTokens tokens = sts.refresh(oauth.getTokens());
+                        session.getClient().setProviderCredentials(new AWSSessionCredentials(tokens.getAccessKey(),
+                                tokens.getSecretAccessKey(), tokens.getSessionToken()));
+                        // Try again
+                        return true;
```

Comment:
I'd need a flow/state diagram 😉

Reply:
Credentials are accessed from `client.getProviderCredentials()` in request interceptors and thus available upon retry of the request.